homework/promise.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+/* cSpell:disable */
+{
+  function fetchJSONWithPromise(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
+      xhr.open('GET', url);
+      xhr.responseType = 'json';
+      xhr.onload = () => {
+        if (xhr.status < 400) {
+          resolve(xhr.response);
+        } else {
+          reject(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+        }
+      };
+      xhr.onerror = () => reject(new Error('Network request failed'));
+      xhr.send();
+    });
+  }
+
+  function createAndAppend(name, parent, options = {}) {
+    const elem = document.createElement(name);
+    parent.appendChild(elem);
+    Object.keys(options).forEach(key => {
+      const value = options[key];
+      if (key === 'text') {
+        elem.textContent = value;
+      } else {
+        elem.setAttribute(key, value);
+      }
+    });
+    return elem;
+  }
+
+  function removeChildren(element) {
+    while (element.firstChild) {
+      element.removeChild(element.firstChild);
+    }
+  }
+
+  function renderError(error) {
+    const root = document.getElementById('root');
+    createAndAppend('div', root, { text: error.message, class: 'alert-error', role: 'alert' });
+  }
+
+  function app(repos) {
+    const root = document.getElementById('root');
+    const container = createAndAppend('div', root, {
+      class: 'container',
+    });
+    const header = createAndAppend('header', container, {
+      class: 'header',
+    });
+    createAndAppend('p', header, {
+      text: 'HYF Repositories',
+    });
+    const select = createAndAppend('select', header, {
+      id: 'list',
+    });
+    for (let i = 0; i < repos.length; i++) {
+      createAndAppend('option', select, { text: repos[i].name, value: i });
+    }
+    const repoContainer = createAndAppend('div', container, { class: 'container-repo' });
```

Comment:
I like the names which you have chosen for variables.

Reply:
Thanks :)